src/ggrc/assets/javascripts/components/access_control_list/access_control_list.js

```diff
@@ -0,0 +1,174 @@
+       * nothing happens.
+       *
+       * @param {CMS.Models.Person} person - the user to grant a role to
+       * @param {Number} roleId - ID if the role to grant
+       */
+      grantRole: function (person, roleId) {
+        var inst = this.instance;
+
+        var roleEntry = _.find(
+          inst.attr('access_control_list'),
+          {person_id: person.id, ac_role_id: roleId}
+        );
+
+        if (roleEntry) {
+          console.warn(
+            'User ', person.id, 'already has role', roleId, 'assigned');
+          return;
+        }
+
+        roleEntry = {person: person, ac_role_id: roleId};
+        inst.attr('access_control_list').push(roleEntry);
+
+        if (!this.autosave) {
+          this.attr('grantingRoleId', 0);
+          this._rebuildRolesInfo();
+          return;
+        }
+
+        inst.save()
+          .done(function () {
+            GGRC.Errors.notifier('success', 'User role added.');
+            this._rebuildRolesInfo();
+          }.bind(this))
+          .fail(function () {
+            GGRC.Errors.notifier('error', 'Adding user role failed.');
+          })
+          .always(function () {
+            this.attr('grantingRoleId', 0);
+          }.bind(this));
+      },
+
+      /**
+       * Revoke role from user on the model instance.
+       *
+       * If the user already does not have this role assigned on the model
+       * instance, nothing happens.
+       *
+       * @param {CMS.Models.Person} person - the user to grant a role to
+       * @param {Number} roleId - ID if the role to grant
+       */
+      revokeRole: function (person, roleId) {
+        var inst = this.instance;
+
+        var idx = _.findIndex(
+          inst.attr('access_control_list'),
+          {person_id: person.id, ac_role_id: roleId}
+        );
+
+        if (idx < 0) {
+          console.warn('Role ', roleId, 'not found for user', person.id);
+          return;
+        }
+
+        inst.access_control_list.splice(idx, 1);
+
+        if (!this.autosave) {
+          this._rebuildRolesInfo();
+          return;
+        }
+
+        inst.save()
+          .done(function () {
+            GGRC.Errors.notifier('success', 'User role removed.');
+            this._rebuildRolesInfo();
+          }.bind(this))
+          .fail(function () {
+            GGRC.Errors.notifier('error', 'Removing user role failed.');
+          });
+      },
+
+      /**
+       * Update the role info object used by the template to display the roles
+       * and the users which have them granted.
+       */
+      _rebuildRolesInfo: function () {
+        var roleAssignments;
+        var roles;
+        var rolesInfo;
+        var instance = this.instance;
+
+        if (!instance) {
+          this.attr('rolesInfo', []);
+          return;
+        }
+
+        roleAssignments = _.groupBy(instance.access_control_list, 'ac_role_id');
+
+        roles = _.filter(
+          GGRC.access_control_roles, {object_type: instance.type});
```

Comment:
Does `GGRC.access_control_roles` ever change in a lifetime of a single page load? If not, we can cache `roles` to avoid unnecessary filtering with the same result every time (b/c `instance.type` does not change either in all practical use cases).